apollo-router/src/spec/query.rs

```diff
@@ -239,31 +229,31 @@ impl Query {
                 }
             }
             Some(Value::Null) => {
-                // Detect if root __typename is asked in the original query (the qp doesn't put root __typename in subselections)
-                // cf https://github.com/apollographql/router/issues/1677
-                let operation_kind_if_root_typename = original_operation.and_then(|op| {
-                    op.selection_set
-                        .iter()
-                        .any(|f| f.is_typename_field())
-                        .then(|| *op.kind())
```

Comment:
Issue with old code:
1. It doesn't handle __typename inside fragments
2. If subgraphql returns `{ data: null, errors: [...] }` due to null propagation, this code still injects the type name, which breaks spec

Reply:
I found correct fix for it:
https://github.com/apollographql/router/blob/cad52650f68ce260e7b1e829fae2ef8959d0a8ea/apollo-router/src/query_planner/execution.rs#L315